File: src/models/beanie/record.py
"""
This module defines data models for representing historical records of vehicle trips,
including trip details, vehicle information, agency data, and associated metadata.
"""

from datetime import datetime

from beanie import Document
from pydantic import BaseModel

from enums import VehicleType


class Route(BaseModel):
    """
    Represents a route record.

    Attributes:
        id (str): Unique identifier for the route.
        name (str): Name of the route.
        route_short_name (str): Short name of the route. (E.g. `401`)
        route_long_name (str | None): Long name of the route. (E.g. `Paralowie to Salisbury`)
        route_type (VehicleType): Type of vehicle used for the route.
    """

    id: str
    short_name: str
    long_name: str | None = None
    type: VehicleType = VehicleType.UNKNOWN


class Trip(BaseModel):
    """
    Represents a trip record.

    Attributes:
        id (str): Unique identifier for the trip.
        direction (int): Direction of the trip.
        route (Route): Route identifier.
    """

    id: str
    direction: int
    route: Route


class Vehicle(BaseModel):
    """
    Represents a vehicle record.

    Attributes:
        id (str): Unique identifier for the vehicle.
        plate (str | None): License plate number of the vehicle, or None if not assigned.
        type (int): Integer representing the type of vehicle.
    """

    id: str
    label: str | None
    plate: str | None
    type: VehicleType = VehicleType.UNKNOWN


class Agency(BaseModel):
    """
    Represents an agency.

    Attributes:
        id (str): Unique identifier for the agency.
        name (str): Name of the agency.
        url (str): URL of the agency's website.
        timezone (str): Timezone of the agency. (In IANA format, e.g. `Australia/Adelaide`)
    """

    id: str
    name: str
    url: str
    timezone: str


class Record(Document):
    """
    Represents a historical record of a vehicle trip for statistical purposes.

    Attributes:
        vehicle (Vehicle): Vehicle information for the record.
        trip (Trip): Trip information for the record.
        agency (Agency): Agency information for the record.
        timestamp (datetime): Timestamp of the record.
    """

    vehicle: Vehicle

    trip: Trip

    agency: Agency

    timestamp: datetime

    class Settings:
        name = "records"

File: internal/models/record.go
// Reusable application models
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
)

type VehicleType int

const (
	VehicleTypeTram      VehicleType = 0
	VehicleTypeSubway    VehicleType = 1
	VehicleTypeRail      VehicleType = 2
	VehicleTypeBus       VehicleType = 3
	VehicleTypeFerry     VehicleType = 4
	VehicleTypeCableTram VehicleType = 5
	VehicleTypeAerial    VehicleType = 6
	VehicleTypeFunicular VehicleType = 7
	VehicleTypeTrolley   VehicleType = 11
	VehicleTypeMonorail  VehicleType = 12
	VehicleTypeUnknown   VehicleType = 1000
)

// Represents a route a vehicle has taken
type Route struct {
	ID        string      `bson:"id" json:"id"`
	ShortName *string     `bson:"short_name,omitempty" json:"short_name,omitempty"`
	LongName  *string     `bson:"long_name,omitempty" json:"long_name,omitempty"`
	Type      VehicleType `bson:"type" json:"type"`
}

// Represents a trip taken by a vehicle
type Trip struct {
	ID        string `bson:"id" json:"id"`
	Direction uint64 `bson:"direction" json:"direction"`
	Route     Route  `bson:"route" json:"route"`
}

// Represents a vehicle
type Vehicle struct {
	ID    string      `bson:"id" json:"id"`
	Label *string     `bson:"label,omitempty" json:"label,omitempty"`
	Plate *string     `bson:"plate,omitempty" json:"plate,omitempty"`
	Type  VehicleType `bson:"type" json:"type"`
}

// Represents an agency that operates a vehicle
type Agency struct {
	ID       string `bson:"id" json:"id"`
	Name     string `bson:"name" json:"name"`
	URL      string `bson:"url" json:"url"`
	Timezone string `bson:"timezone" json:"timezone"`
}

// Represents a unique trip that a vehicle has taken
type Record struct {
	ID        bson.ObjectID `bson:"_id,omitempty" json:"id,omitempty"`
	Vehicle   Vehicle       `bson:"vehicle" json:"vehicle"`
	Trip      Trip          `bson:"trip" json:"trip"`
	Agency    Agency        `bson:"agency" json:"agency"`
	Timestamp time.Time     `bson:"timestamp" json:"timestamp"`
}
